### PR TITLE
Fix ModelCheckpoint file_exists OOM in DDP

### LINKED ISSUE #19674: CUDA memory increase (caused CUDA OOM) when saving checkpoint at the `train_epoch_end`
### Bug description

We have monitored that CUDA memory increase when saving checkpoint at the `on_train_epoch_end` (It caused the CUDA OOM). Instead of directly give a OOM at the beginning of training step, this behavior caused by automatically checkpoint mechanism. 

We have checked any possible place may cause memory leak in training step, and CUDA memory didn't increase during the training step accumelately.

The following figure show the error message :
image

---

We found that when we set the training data size smaller, the error will not caused in first epoch. When we set the training data size bigger, then the error will occurs at the same place.

```
# Dataset code snippet
from pathlib import Path
import webdataset as wds
import pytorch_lightning as pl
from torch.utils.data import DataLoader
import torch

import io
import imageio
import braceexpand

from torch.distributed import get_rank, get_world_size
from torch.utils.data import get_worker_info, DistributedSampler
import json 

class Iter_ds(torch.utils.data.IterableDataset):

    def __init__(self, urls, trfs, n_sample):
        self.urls = urls
        self.trfs = trfs
        self.n_sample = n_sample
        
    def __len__(self):
        return self.n_sample // get_world_size()
                       
    def __iter__(self):
        process_rank = get_rank()
        world_size = get_world_size()
        for url in self.urls:  
            # feed dummy pipe `wds.split_by_worker`, which is the default of workersplit, thus it'll not affect the results
            dataset = wds.WebDataset(url, nodesplitter=None).shuffle(1000).decode('pilrgb', handler=wds.warn_and_continue).to_tuple("jpg", "txt", handler=wds.warn_and_continue).map(self.trfs)
            for batch_id, sample in enumerate(dataset):
                # assign a independent batch for the gpu wrt. gpu_id (nodesplitter in here)
                if batch_id % world_size == process_rank: 
                    yield sample
                # skip the batch it doesn't belong to the gpu
                else: 
                    continue

class Liaon115M(pl.LightningDataModule):

    def __init__(self, 
        data_dir: str = "path/to/dir", 
        split_ratio: list = [0.7, 0.1, 0.2], 
        img_transforms = None, 
        txt_transforms = None, 
        num_workers: int = 4, 
        batch_size: int = 16,
        num_epoch: int = 1,
        pin_memory: bool = False
    ):
        super().__init__()
        self.data_dir = Path(data_dir)
        self.split_ratio = split_ratio
        self.img_transforms = img_transforms
        self.txt_transforms = txt_transforms

        self.trfs = lambda tup : (self.img_transforms(tup[0]), self.txt_transforms(tup[1]))

        self.batch_size = batch_size
        self.num_epoch = num_epoch
        self.num_workers = num_workers
        self.pin_memory = pin_memory

        # laion 115M, success downloaded 104881705 records.. 
        # we setup last 5 digit as mask for 'un-equal tar file issue'
        #total_num_ds, mask  = 1048_81705, 81705
        
    # chking dataset is downloaded..
    def prepare_data(self): 
        with self.data_dir.open('r') as js_ptr:
            self.tar_dict = json.load(js_ptr)
        
        tar_lst = list(self.tar_dict.keys())
        n_shard = len(tar_lst)
        
        tra_ratio, val_ratio, _ = self.split_ratio
        self.tra_lst = tra_lst = tar_lst[ : int(n_shard * tra_ratio) ]
        self.val_lst = val_lst = tar_lst[ len(tra_lst) : len(tra_lst) + int(n_shard * val_ratio) ]
        self.tst_lst = tar_lst[ len(tra_lst) + len(val_lst) :  ]

    def _get_sample_num(self, tar_lst):
            cnt = 0
            for tar_key in tar_lst:
                cnt += self.tar_dict[tar_key]
            return cnt

    def setup(self, stage: str = 'train', rank=None, world_size=1):
        self.prepare_data()
        if stage == 'train':
            n_tra_sample = self._get_sample_num(self.tra_lst)
            print(f'Total sample in train ds : {n_tra_sample}')
            self.laion_train = Iter_ds(self.tra_lst, trfs=self.trfs, n_sample=n_tra_sample)
            n_val_sample = self._get_sample_num(self.val_lst)
            print(f'Total sample in valid ds : {n_val_sample}')
            self.laion_valid = Iter_ds(self.val_lst, trfs=self.trfs, n_sample=n_val_sample)
        else:
            n_tst_sample = self._get_sample_num(self.tst_lst)
            print(f'Total sample in test ds : {n_tst_sample}')
            self.laion_test = Iter_ds(self.tst_lst, trfs=self.trfs, n_sample=n_tst_sample)

    def train_dataloader(self):
        loader = torch.utils.data.DataLoader(
            self.laion_train, 
            batch_size=self.batch_size, 
            shuffle=False, 
            pin_memory=True, 
            num_workers=self.num_workers,
            prefetch_factor=2,
            drop_last=True
        )
        return loader 

    def val_dataloader(self):
        loader = torch.utils.data.DataLoader(
            self.laion_valid, 
            batch_size=self.batch_size, 
            shuffle=False, 
            pin_memory=True, 
            num_workers=self.num_workers,
            prefetch_factor=2,
            drop_last=True
        )
        return loader 
       
    def test_dataloader(self):
        return DataLoader(self.laion_test, batch_size=self.batch_size, num_workers=self.num_workers, pin_memory=self.pin_memory)

    def predict_dataloader(self):
        return DataLoader(self.laion_test, batch_size=self.batch_size, num_workers=self.num_workers, pin_memory=self.pin_memory)


``` 

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

```python
Sorry, but we have no idea how to write a minimum example to reproduce the bug. It's not easy while the code is structured.
```

### Error messages and logs

```
# Error messages and logs here please

Using 16bit Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
You are using a CUDA device ('NVIDIA A100-SXM4-80GB') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precis
i
ion('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torc[194/785]
at32_matmul_precision.html#torch.set_float32_matmul_precision
[rank: 0] Seed set to 42
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/2
[rank: 1] Seed set to 42
Total sample in train ds : 309617
Total sample in valid ds : 4588
[rank: 1] Seed set to 42
Initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/2
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 2 processes
----------------------------------------------------------------------------------------------------

Total sample in test ds : 53430531
Total sample in test ds : 53430531
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]

  | Name     | Type     | Params                                                                                                             [175/785]
--------------------------------------
0 | model    | Glow_T2I | 23.1 M
1 | clip_mod | CLIP     | 151 M
--------------------------------------
23.1 M    Trainable params
151 M     Non-trainable params
174 M     Total params
697.404   Total estimated model params size (MB)
Epoch 0:  48%|▍| 2327/4837 [48:22<52:10,  0.80it/s, v_num=l29k, train_txt_sim_loss=1.560, train_img_sim_loss=0.078, log_p=-2.86, logdet=4.810, train_g
low_loss=-Epoch 0:  49%|▍| 2386/4837 [49:35<50:56,  0.80it/s, v_num=l29k, train_txt_sim_loss=1.550, train_img_sim_loss=0.109, log_p=-2.46, logdet=4.81
0, train_glow_loss=-Epoch Epoch 0:  97%|▉| 4677/4837 [1:37:07<03:19,  0.80it/s, v_num=l29k, train_txt_sim_loss=1.530, train_img_sim_loss=0.0767, log_p
=-1.65, logdet=4.370, train_glEpoch 0:  97%|▉| 4678/4837 [1:37:09<03:18,  0.80it/s, v_num=l29k, train_txt_sim_loss=1.530, train_img_sim_loss=0.0767, l
og_p=-1.65, logdet=4.370, train_glow_losEpoch 0: 100%|▉| 4836/4837 [1:40:25<00:01,  0.80it/s, v_num=l29k, train_txt_sim_loss=1.560, train_img_sim_loss
=0.0717, log_p=-1.57, logdet=4.400, train_glow_losTraceback (most recent call last):
  File "/data/harry/CAR-GANV2/train.py", line 59, in <module>
    main(cfger)
  File "/data/harry/CAR-GANV2/train.py", line 46, in main
    trainer.fit(train_loop, data_module)

File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 544, in fit                              [156/785]
    call._call_and_handle_interrupt(
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/trainer/call.py", line 43, in _call_and_handle_interrupt
    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/strategies/launchers/subprocess_script.py", line 105, in launch
    return function(*args, **kwargs)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 580, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 987, in _run
    results = self._run_stage()
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1033, in _run_stage
    self.fit_loop.run()
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 206, in run
    self.on_advance_end()
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 378, in on_advance_end
    call._call_callback_hooks(trainer, "on_train_epoch_end", monitoring_callbacks=True)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/trainer/call.py", line 208, in _call_callback_hooks
    fn(trainer, trainer.lightning_module, *args, **kwargs)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 324, in on_train_epoch_end
    self._save_topk_checkpoint(trainer, monitor_candidates)

self._save_none_monitor_checkpoint(trainer, monitor_candidates)                                                                          [135/785]
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 711, in _save_none_monitor_checkpoint
    filepath = self._get_metric_interpolated_filepath_name(monitor_candidates, trainer, self.best_model_path)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 662, in _get_metric_interpolated_filepa
th_name
    while self.file_exists(filepath, trainer) and filepath != del_filepath:
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 775, in file_exists
    return trainer.strategy.broadcast(exists)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/pytorch_lightning/strategies/ddp.py", line 307, in broadcast
    torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/torch/distributed/c10d_logger.py", line 72, in wrapper
    return func(*args, **kwargs)
  File "/anaconda/envs/ignf/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 2422, in broadcast_object_list
    object_tensor = torch.empty(  # type: ignore[call-overload]
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 67372036.02 GiB. GPU 1 has a total capacity of 79.18 GiB of which 75.04 GiB is free
. Process 24697 has 4.14 GiB memory in use. Of the allocated memory 1.06 GiB is allocated by PyTorch, and 1.57 GiB is reserved by PyTorch but unalloca
ted. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentati
on for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
[rank: 1] Child process with PID 8701 terminated with code 1. Forcefully terminating all other processes to avoid zombies 🧟
./scripts/train_ignf.sh: line 3:  8186 Killed                  python train.py
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA A100-SXM4-80GB
                - NVIDIA A100-SXM4-80GB
        - available:         True
        - version:           11.8
* Lightning:
        - lightning-utilities: 0.10.1
        - open-clip-torch:   2.24.0
        - pytorch-lightning: 2.2.1
        - torch:             2.2.0+cu118
        - torchaudio:        2.2.0+cu118
        - torchdata:         0.7.1
        - torchmetrics:      1.3.1
        - torchvision:       0.17.0+cu118
* Packages:
        - aiohttp:           3.9.3
        - aiosignal:         1.3.1
        - albumentations:    1.4.1
        - appdirs:           1.4.4
        - asttokens:         2.4.1
        - async-timeout:     4.0.3
        - attrs:             23.2.0
        - braceexpand:       0.1.7
        - certifi:           2024.2.2
        - charset-normalizer: 3.3.2
        - click:             8.1.7
        - comm:              0.2.2
        - contourpy:         1.2.0
        - cycler:            0.12.1
        - dataclasses:       0.6
        - debugpy:           1.6.7
        - decorator:         5.1.1
        - docker-pycreds:    0.4.0
        - easy-configer:     2.2.0
        - einops:            0.7.0
        - exceptiongroup:    1.2.0
        - executing:         2.0.1
        - exifread-nocycle:  3.0.1
        - filelock:          3.13.1
        - fire:              0.5.0
        - fonttools:         4.50.0
        - frozenlist:        1.4.1
        - fsspec:            2024.2.0
        - ftfy:              6.1.3
        - gitdb:             4.0.11
        - gitpython:         3.1.42
        - huggingface-hub:   0.21.3
        - idna:              3.6
        - imageio:           2.34.0
        - img2dataset:       1.45.0
        - importlib-metadata: 7.0.2
        - importlib-resources: 6.3.1
        - ipykernel:         6.29.3
        - ipython:           8.18.1
        - jedi:              0.19.1
        - jinja2:            3.1.3
        - joblib:            1.3.2
        - jupyter-client:    8.6.1
        - jupyter-core:      5.7.2
        - kiwisolver:        1.4.5
        - lazy-loader:       0.3
        - lightning-utilities: 0.10.1
        - markupsafe:        2.1.5
        - matplotlib:        3.8.3
        - matplotlib-inline: 0.1.6
        - memcnn:            1.5.2
        - mpmath:            1.3.0
        - multidict:         6.0.5
        - nest-asyncio:      1.6.0
        - networkx:          3.2.1
        - numpy:             1.26.4
        - nvidia-cublas-cu11: 11.11.3.6
        - nvidia-cublas-cu12: 12.1.3.1
        - nvidia-cuda-cupti-cu11: 11.8.87
        - nvidia-cuda-cupti-cu12: 12.1.105
        - nvidia-cuda-nvrtc-cu11: 11.8.89
        - nvidia-cuda-nvrtc-cu12: 12.1.105
        - nvidia-cuda-runtime-cu11: 11.8.89
        - nvidia-cuda-runtime-cu12: 12.1.105
        - nvidia-cudnn-cu11: 8.7.0.84
        - nvidia-cudnn-cu12: 8.9.2.26
        - nvidia-cufft-cu11: 10.9.0.58
        - nvidia-cufft-cu12: 11.0.2.54
        - nvidia-curand-cu11: 10.3.0.86
        - nvidia-curand-cu12: 10.3.2.106
        - nvidia-cusolver-cu11: 11.4.1.48
        - nvidia-cusolver-cu12: 11.4.5.107
        - nvidia-cusparse-cu11: 11.7.5.86
        - nvidia-cusparse-cu12: 12.1.0.106
        - nvidia-nccl-cu11:  2.19.3
        - nvidia-nccl-cu12:  2.19.3
        - nvidia-nvjitlink-cu12: 12.3.101
        - nvidia-nvtx-cu11:  11.8.86
        - nvidia-nvtx-cu12:  12.1.105
        - open-clip-torch:   2.24.0
        - opencv-python-headless: 4.9.0.80
        - packaging:         23.2
        - pandas:            2.2.1
        - parso:             0.8.3
        - pathlib2:          2.3.7.post1
        - pexpect:           4.9.0
        - pickleshare:       0.7.5
        - pillow:            10.2.0
        - pip:               23.3.1
        - platformdirs:      4.2.0
        - prompt-toolkit:    3.0.42
        - protobuf:          4.25.3
        - psutil:            5.9.8
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - pyarrow:           15.0.0
        - pygments:          2.17.2
        - pyparsing:         3.1.2
        - python-dateutil:   2.9.0
        - pytorch-lightning: 2.2.1
        - pytz:              2024.1
        - pyyaml:            6.0.1
        - pyzmq:             25.1.2
        - regex:             2023.12.25
        - requests:          2.31.0
        - safetensors:       0.4.2
        - scikit-image:      0.22.0
        - scikit-learn:      1.4.1.post1
        - scipy:             1.12.0
        - sentencepiece:     0.2.0
        - sentry-sdk:        1.40.6
        - setproctitle:      1.3.3
        - setuptools:        68.2.2
        - simpleitk:         2.3.1
        - six:               1.16.0
        - smmap:             5.0.1
        - stack-data:        0.6.2
        - sympy:             1.12
        - termcolor:         2.4.0
        - threadpoolctl:     3.3.0
        - tifffile:          2024.2.12
        - timm:              0.9.16
        - torch:             2.2.0+cu118
        - torchaudio:        2.2.0+cu118
        - torchdata:         0.7.1
        - torchmetrics:      1.3.1
        - torchvision:       0.17.0+cu118
        - tornado:           6.4
        - tqdm:              4.66.2
        - traitlets:         5.14.2
        - triton:            2.2.0
        - typing-extensions: 4.10.0
        - tzdata:            2024.1
        - urllib3:           2.2.1
        - wandb:             0.16.3
        - wcwidth:           0.2.13
        - webdataset:        0.2.88
        - wheel:             0.41.2
        - yarl:              1.9.4
        - zipp:              3.17.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.18
        - release:           5.4.0-1109-azure
        - version:           #115~18.04.1-Ubuntu SMP Mon May 22 20:06:37 UTC 2023

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by littlebullGit**
## What does this PR do?

- Use [strategy.reduce_boolean_decision] instead of [broadcast] in [ModelCheckpoint.file_exists].
- Ensure only global rank 0 touches the filesystem when checking for existing checkpoints.
- Avoid `broadcast_object_list` for a simple boolean in DDP to reduce memory pressure in the checkpoint path.
- Add a small DDP test with `monitor=None` to exercise this path.

Fixes #19674

### Motivation and context

In DDP, [strategy.broadcast] is implemented via `torch.distributed.broadcast_object_list`, which serializes the Python object and can allocate unnecessary GPU memory even for a single boolean. For the “file exists” decision we only need a tiny boolean reduction, so [reduce_boolean_decision] is a better fit and addresses the CUDA OOM reported in #19674 while preserving behavior.

### Dependencies

- No new runtime dependencies introduced by this PR.
- Tests rely on `pytorch_lightning_enterprise` being available, as required by [tests/tests_pytorch/conftest.py].

### Tests

All run inside the project `.venv`:

- `python -m pytest tests/tests_pytorch/checkpointing/test_checkpoint_callback_frequency.py`
- `python -m pytest tests/tests_pytorch/checkpointing -k "not legacy_checkpoints"`
- `python -m pytest tests/tests_pytorch/callbacks/test_model_checkpoint_*.py tests/tests_pytorch/trainer/test_trainer.py`

----
📚 Documentation preview 📚: https://pytorch-lightning--21380.org.readthedocs.build/en/21380/

### CODE REVIEW

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -999,8 +999,14 @@ def to_yaml(self, filepath: Optional[_PATH] = None) -> None:
     def file_exists(self, filepath: _PATH, trainer: "pl.Trainer") -> bool:
         """Checks if a file exists on rank 0 and broadcasts the result to all other ranks, preventing the internal
         state to diverge between ranks."""
-        exists = self._fs.exists(filepath)
-        return trainer.strategy.broadcast(exists)
+        # Single-process or strategies without distributed world size: no need for coordination
+        if trainer.world_size == 1:
+            return self._fs.exists(filepath)
```
Comment (justusschock) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]: in these cases the `reduce_boolean_decision` should be a no-op, so could we just do it always for consistency

Comment (littlebullGit) [reply src/lightning/pytorch/callbacks/model_checkpoint.py]: Great point. updated.

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -999,8 +999,10 @@ def to_yaml(self, filepath: Optional[_PATH] = None) -> None:
     def file_exists(self, filepath: _PATH, trainer: "pl.Trainer") -> bool:
         """Checks if a file exists on rank 0 and broadcasts the result to all other ranks, preventing the internal
```
Comment (bhimrazy) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]: Should we also update the docstring here, then? It still mentions "broadcasts" but the implementation now uses "reduce". 

While technically correct, it might be clearer to say "synchronizes the result across all ranks" to avoid confusion with the terminology.
wdyt @littlebullGit @justusschock ?

Comment (justusschock) [inline src/lightning/pytorch/callbacks/model_checkpoint.py]:
```suggestion
        """Checks if a file exists on rank 0 and synchronizes the result to all other ranks, preventing the internal
```
